service_grpc_flight/src/lib.rs

```diff
@@ -134,19 +155,32 @@ impl ReadInfo {
     fn decode_json(ticket: &[u8]) -> Result<Self> {
         let json_str = String::from_utf8(ticket.to_vec()).context(InvalidJsonTicketSnafu {})?;
 
-        let read_info: ReadInfo =
+        #[derive(Deserialize, Debug)]
+        struct ReadInfoJson {
+            namespace_name: String,
+            sql_query: String,
+        }
+
+        let read_info: ReadInfoJson =
             serde_json::from_str(&json_str).context(InvalidQuerySnafu { query: &json_str })?;
 
-        Ok(read_info)
+        Ok(Self {
+            namespace_name: read_info.namespace_name,
+            query: Query::Sql(read_info.sql_query), // JSON is always SQL
+        })
     }
 
     fn decode_protobuf(ticket: &[u8]) -> Result<Self> {
         let read_info =
             proto::ReadInfo::decode(Bytes::from(ticket.to_vec())).context(InvalidTicketSnafu {})?;
 
         Ok(Self {
-            namespace_name: read_info.namespace_name,
-            sql_query: read_info.sql_query,
+            namespace_name: read_info.namespace_name.clone(),
+            query: match read_info.query_type {
+                0 /* unspecified */ | 1 /* SQL */ => Query::Sql(read_info.sql_query),
+                2 => Query::InfluxQL(read_info.sql_query),
+                _ => return Err(Error::InvalidQueryType),
+            },
```

Comment:
There is a function query_type() that is strongly typed -- I recommend you could use

```suggestion
            query: match read_info.query_type() {
                QueryType::Unspecified | QueryType::Sql => Query::Sql(read_info.sql_query),
                QueryType::InfluxQl  => Query::InfluxQL(read_info.sql_query),
            },
```

Reply:
Ahh yes, thank you! Cleans up the code nicely too, and removes the possibility of an error.